### DISCUSSION TITLE
[RESOLVED] ReducerProtocol: Missing ViewStore updates in Release in Xcode 14

### DISCUSSION THREAD

**Discussion by PorterHoskins** [pointfreeco/swift-composable-architecture#1379] [Beta]
Loving the beta. It really is great. I put up a TestFlight build to try it out and my entire app was broken. After looking into it more, it seems like there are `ViewStore` updates that are not getting fired on this branch when the app is built in release mode in Xcode 14. 

It doesn't seem to be an issue directly with `ReducerProtocol` and happens with both reducer paradigms. I tried turning off Whole Module Optimization and Speed Optimization build settings. I tried converting everything to the new concurrency model. I got rid of all the deprecated build warnings. Everything is straight TCA with no custom extensions. 

0.40.1 works fine from Xcode 14 in release mode. The beta works fine in debug and in release on Xcode 13. I tried to reproduce the issue in a small sample project but I can't get it to reproduce. Are you seeing any issues like this with Isowords by chance?

**Comment by jaanussiim** [comment]
This reminds me. While trying to create a sample project for another issue got to this issue where actions disappear in release build.

You should be able to copy it into view controller of default new app project. And adding protocol-beta + xctest-dynamic-overlay to Packages.

```
import ComposableArchitecture
import Foundation
import SwiftUI
import UIKit
import XCTestDynamicOverlay


internal class ViewController: UIViewController {
    private lazy var client = TimeEntriesClient(
        onLoadEntries: {
            _ in
                
            try await DispatchQueue.main.sleep(for: .milliseconds(500))
            
            return [
                Entry(id: Int64.random(in: 0...Int64.max)),
                Entry(id: Int64.random(in: 0...Int64.max)),
                Entry(id: Int64.random(in: 0...Int64.max)),
                Entry(id: Int64.random(in: 0...Int64.max)),
                Entry(id: Int64.random(in: 0...Int64.max)),
                Entry(id: Int64.random(in: 0...Int64.max))
            ]
        }
    )
    
    private lazy var store = Store(
        initialState: Application.State(),
        reducer: Application()
            .dependency(\.timeEntriesClient, client)
    )
    private lazy var appView = ApplicationView(store: store)
    private lazy var hosting = UIHostingController(rootView: appView)
    
    override func viewDidLoad() {
        super.viewDidLoad()

        addChild(hosting)
        view.addSubview(hosting.view)
        hosting.view.translatesAutoresizingMaskIntoConstraints = false
        hosting.view.leadingAnchor.constraint(equalTo: view.leadingAnchor).isActive = true
        hosting.view.topAnchor.constraint(equalTo: view.topAnchor).isActive = true
        hosting.view.trailingAnchor.constraint(equalTo: view.trailingAnchor).isActive = true
        hosting.view.bottomAnchor.constraint(equalTo: view.bottomAnchor).isActive = true
    }
}

public struct TimeEntriesClient {
    internal let onLoadEntries: ((Date) async throws -> [Entry])
}

extension TimeEntriesClient {
    public static let unimplemented = TimeEntriesClient(
        onLoadEntries: XCTUnimplemented("\(Self.self).onLoadEntries")
    )
}

private enum TimeEntriesClientKey: DependencyKey {
    static var liveValue = TimeEntriesClient.unimplemented
    static let previewValue = TimeEntriesClient.unimplemented
    static let testValue = TimeEntriesClient.unimplemented
}

extension DependencyValues {
    public var timeEntriesClient: TimeEntriesClient {
        get { self[TimeEntriesClientKey.self] }
        set { self[TimeEntriesClientKey.self] = newValue }
    }
}


public struct Application: ReducerProtocol {
    public struct State: Equatable {
        internal var timeEntriesState = TimeEntries.State()
    }
    
    public enum Action {
        case timeEntries(TimeEntries.Action)
    }
    
    public var body: some ReducerProtocol<State, Action> {
        Reduce {
            state, action in
        
            switch action {
            case .timeEntries:
                return .none
            }
        }
        Scope(state: \.timeEntriesState, action: /Action.timeEntries) {
            TimeEntries()
        }
    }
}

public struct Entry: Equatable {
    let id: Int64
}

public struct TimeEntry: ReducerProtocol {
    public struct State: Equatable, Identifiable {
        public var id: Int64 {
            entry.id
        }
        
        let entry: Entry
    }
    
    public enum Action {
        case tapped
    }
    
    public var body: some ReducerProtocol<State, Action> {
        Reduce {
            state, action in
            
            switch action {
            case .tapped:
                return .none
            }
        }
    }
}

public struct TimeEntries: ReducerProtocol {
    public struct State: Equatable {
        internal var entries = IdentifiedArrayOf<TimeEntry.State>()
        @BindableState var activeDate = Date.now
        @BindableState var overlayDate = Date.now
        internal var calendarShown = false
    }
    
    public enum Action: BindableAction {
        case loadEntries
        case loaded(TaskResult<[Entry]>)
        case toggleCalendar
        
        
        case binding(BindingAction<TimeEntries.State>)
        case timeEntry(id: TimeEntry.State.ID, action: TimeEntry.Action)
    }
    
    @Dependency(\.timeEntriesClient) var timeEntriesClient
    
    public var body: some ReducerProtocol<State, Action> {
        BindingReducer()
        Reduce {
            state, action in
            
            switch action {
            case .binding(\.$activeDate):
                return Effect(value: .loadEntries)
                
            case .loadEntries:
                return Effect.task {
                    [date = state.activeDate] in
                    
                    await .loaded(
                        TaskResult {
                            try await timeEntriesClient.onLoadEntries(date)
                        }
                    )
                }
                
            case .loaded(.failure):
                return .none
                
            case .loaded(.success(let entries)):
                dump(entries)
                state.entries = IdentifiedArrayOf(uniqueElements: entries.map(TimeEntry.State.init(entry:)))
                return .none
                
            case .toggleCalendar:
                state.calendarShown.toggle()
                return .none
                
            case .binding:
                return .none
                
            case .timeEntry:
                return .none
            }
        }
        .forEach(\.entries, action: /Action.timeEntry(id:action:)) {
            TimeEntry()
        }
        .debug()
    }
}

public struct TimeEntriesView: View {
    internal let store: StoreOf<TimeEntries>
    
    public var body: some View {
        WithViewStore(store) {
            viewStore in
            
            NavigationView {
                ZStack {
                    VStack {
                        DatePicker(
                            "Entries on:",
                            selection: viewStore.binding(\.$activeDate), displayedComponents: .date
                        )
                        .padding(.horizontal)
                        List {
                            ForEachStore(store.scope(state: \.entries, action: TimeEntries.Action.timeEntry(id:action:))) {
                                entryStore in
                                
                                WithViewStore(entryStore) {
                                    entryViewStore in
                                    
                                    Text(String(describing: entryViewStore.id))
                                }
                            }
                        }
                        .listStyle(PlainListStyle())
                    }
                    
                    if viewStore.calendarShown {
                        Color.white.opacity(1)
                        VStack {
                            DatePicker(
                                "Entries on:",
                                selection: viewStore.binding(\.$overlayDate), displayedComponents: .date
                            )
                            .datePickerStyle(.graphical)
                            .padding(.horizontal)
                            Spacer()
                        }
                    }
                    
                }
                .toolbar {
                    ToolbarItem {
                        Button(action: { viewStore.send(.toggleCalendar) }) {
                            Image(systemName: "calendar")
                        }
                    }
                }
            }
            .navigationViewStyle(StackNavigationViewStyle())
            .onAppear {
                viewStore.send(.loadEntries)
            }

        }
    }
}

public struct ApplicationView: View {
    internal let store: StoreOf<Application>
    
    public var body: some View {
        WithViewStore(store) {
            viewStore in
            
            TabView {
                TimeEntriesView(store: store.scope(state: \.timeEntriesState, action: Application.Action.timeEntries))
                    .tabItem {
                        Label("Time entries", systemImage: "rectangle.stack")
                    }
            }
        }
    }
}

```

**Comment by mbrandonw** [comment]
Thanks @PorterHoskins for bringing this up and thanks @jaanussiim for giving us a small repro! We can see this behavior when running locally, but we haven't figured out what exactly is going on yet. We're both busy for the next few days with some travel, but we will look into it soon!

In the meantime, if you have time @jaanussiim, would you mind providing us with a port of your code sample that runs on main? That will give us something to hop back and forth between to see what is breaking down. If you don't have time it's OK, we can do it ourselves later.

**Comment by jaanussiim** [comment]
@mbrandonw - by "runs on main" you mean `@main struct SwiftUICheckApp: App`? :)

```
import ComposableArchitecture
import SwiftUI
import XCTestDynamicOverlay

@main struct SwiftUICheckApp: App {
    private let appStore = Store(initialState: Application.State(), reducer: Application())
    private let entriesStore = Store(initialState: TimeEntries.State(), reducer: TimeEntries())
    
    var body: some Scene {
        WindowGroup {
            TabView {
                TimeEntriesView(
                    store: appStore.scope(state: \.timeEntriesState, action: Application.Action.timeEntries)
                )
                .tabItem {
                    Label("Scoped", systemImage: "star.fill")
                }
                TimeEntriesView(
                    store: entriesStore
                )
                .tabItem {
                    Label("Direct", systemImage: "wand.and.stars")
                }
            }
        }
    }
}

public struct Application: ReducerProtocol {
    public struct State: Equatable {
        internal var timeEntriesState = TimeEntries.State()
    }
    
    public enum Action {
        case timeEntries(TimeEntries.Action)
    }
    
    public var body: some ReducerProtocol<State, Action> {
        Reduce {
            state, action in
        
            switch action {
            case .timeEntries:
                return .none
            }
        }
        Scope(state: \.timeEntriesState, action: /Action.timeEntries) {
            TimeEntries()
        }
    }
}

public struct Entry: Equatable {
    let id: Int64
}

public struct TimeEntries: ReducerProtocol {
    public struct State: Equatable {
        @BindableState var activeDate = Date.now
    }
    
    public enum Action: BindableAction {
        case loadEntries
        
        // ######### comment out
        case loaded(TaskResult<[Entry]>)
        // #########
        
        case binding(BindingAction<TimeEntries.State>)
    }
    
    public var body: some ReducerProtocol<State, Action> {
        BindingReducer()
        Reduce {
            state, action in
            
            switch action {
            case .binding(\.$activeDate):
                return Effect(value: .loadEntries)
                
            case .loadEntries:
                return .none

            // ######### comment out
            case .loaded(let result):
                return .none
            // #########
                                
            case .binding:
                return .none
            }
        }
        .debug()
    }
}

public struct TimeEntriesView: View {
    internal let store: StoreOf<TimeEntries>
    
    public var body: some View {
        WithViewStore(store) {
            viewStore in
            
            NavigationView {
                ZStack {
                    VStack {
                        DatePicker(
                            "Entries on:",
                            selection: viewStore.binding(\.$activeDate), displayedComponents: .date
                        )
                        .padding(.horizontal)
                        Spacer()
                        Text(viewStore.activeDate, style: .date)
                    }
                }
            }

        }
    }
}
```


I believe I got the sample down to bare minimum. Only one binding action. First tab uses scoped store and actions do not work in release mode. In second tab, using direct store, actions work.

If you comment out `TimeEntries.Action.loaded(TaskResult<[Entry]>)` then the binding action also fires on the first tab

**Reply by PorterHoskins** [reply]
I think they mean could you provide the same sample code but using the old reducer paradigm. Like no ReducerProtocol.

**Comment by jaanussiim** [comment]
Anyway - here is a version without ReducerProtocol. Has the same behaviour in main branch. No actions in release mode

```
import ComposableArchitecture
import SwiftUI

@main
struct MainBranchCheckApp: App {
    private let appStore = Store(
        initialState: Application.State(),
        reducer: Application.reducer,
        environment: ()
    )
    private let entriesStore = Store(
        initialState: TimeEntries.State(),
        reducer: TimeEntries.reducer,
        environment: ()
    )
    
    var body: some Scene {
        WindowGroup {
            TabView {
                TimeEntriesView(
                    store: appStore.scope(state: \.timeEntriesState, action: Application.Action.timeEntries)
                )
                .tabItem {
                    Label("Scoped", systemImage: "star.fill")
                }
                TimeEntriesView(
                    store: entriesStore
                )
                .tabItem {
                    Label("Direct", systemImage: "wand.and.stars")
                }
            }
        }
    }
}

public struct Application {
    public struct State: Equatable {
        internal var timeEntriesState = TimeEntries.State()
    }
    
    public enum Action {
        case timeEntries(TimeEntries.Action)
    }
    
    public static let reducer = Reducer<Application.State, Application.Action, Void>.combine(
        TimeEntries.reducer
            .pullback(
                state: \.timeEntriesState,
                action: /Action.timeEntries,
                environment: { $0 }
            ),
        core
    )
    
    
    private static let core = Reducer<Application.State, Application.Action, Void> {
        state, action, env in
        
        switch action {
        case .timeEntries:
            return .none
        }
    }
}

public struct Entry: Equatable {
    let id: Int64
}

public struct TimeEntries {
    public struct State: Equatable {
        @BindableState var activeDate = Date.now
    }
    
    public enum Action: BindableAction {
        case loadEntries
        
        // ######### comment out
        case loaded(TaskResult<[Entry]>)
        // #########
        
        case binding(BindingAction<TimeEntries.State>)
    }
    
    public static let reducer = Reducer<TimeEntries.State, TimeEntries.Action, Void> {
        state, action, env in
        
        switch action {
        case .binding(\.$activeDate):
            return Effect(value: .loadEntries)
            
        case .loadEntries:
            return .none

        // ######### comment out
        case .loaded(let result):
            return .none
        // #########
                            
        case .binding:
            return .none
        }
    }
    .binding()
}

public struct TimeEntriesView: View {
    internal let store: Store<TimeEntries.State, TimeEntries.Action>
    
    public var body: some View {
        WithViewStore(store) {
            viewStore in
            
            NavigationView {
                ZStack {
                    VStack {
                        DatePicker(
                            "Entries on:",
                            selection: viewStore.binding(\.$activeDate), displayedComponents: .date
                        )
                        .padding(.horizontal)
                        Spacer()
                        Text(viewStore.activeDate, style: .date)
                    }
                }
            }

        }
    }
}

```

**Reply by mbrandonw** [reply]
Oh sorry, I thought you said your code worked on `main` and only broke on the `protocol-beta` branch, but I see that's not the case. If this is happening on main then it seems like a bigger problem.

**Reply by tgrapperon** [reply]
My gut feeling is that the `Any` in `TaskResult` is making the `CasePath` hit this: https://github.com/pointfreeco/swift-case-paths/issues/71. I'll try to check soon.

**Reply by tgrapperon** [reply]
I don't have the time to dig deeper and confirm that it is the same issue, but making the  adding another `.noop` action to `MyApplication.Action` or making `TimeEntries.Action` `indirect` seems to fix the issue here too. So it looks like the layout of the enum is different for release builds and it fails to extract the action.

**Reply by PorterHoskins** [reply]
These workarounds don't work for me. I am also never using an enum with a single value. 

What I see is the reducer processes the value correctly and the state is updated but the ViewStore is not updated. If I force a re-render in SwiftUI then the state is correct.

**Reply by PorterHoskins** [reply]
@mbrandonw I can confirm that `main` behaves appropriately and the ViewStore publishes updates.

**Reply by tgrapperon** [reply]
@PorterHoskins, can you make sure that you have the latest version of `protocol-beta`, where `WithViewStore.swift` should have `@ObservedObject private var viewStore: ViewStore<ViewState, ViewAction>` at line 120?. If not, you can try to update `swift-composable-architecture` package by right clicking in the package list at the bottom of the left panel in Xcode.

**Reply by PorterHoskins** [reply]
Yeah I have the latest 👍

**Reply by tgrapperon** [reply]
This is weird because `ViewStore` is mostly unchanged between the two branches, and `WithViewStore` the same. The main difference I can think about is the `ScopedReducer` that is maybe not forwarding to the publisher correctly in Release.
Would you mind trying this [branch](https://github.com/tgrapperon/swift-composable-architecture/tree/protocol-beta-check) where I force the store to use the same approach as `main`? You can simply replace `pointfreeco` by `tgrapperon` and `protocol-beta` by `protocol-beta-check`. This would likely allow to narrow the problem by a lot if it works.

**Reply by PorterHoskins** [reply]
That does indeed fix the issue

**Reply by erikolsson** [reply]
Just chiming in that after debugging some strange behavior that appeared in iOS16, I finally found that I had wrapped `NavigationView` in `WithViewStore`, (similar to your code above).

My issue was fixed by changing:
```
WithViewStore(store) { viewStore in
      NavigationView {
          /* MyView */
      }
}
```
Into:
```
NavigationView {
      WithViewStore(store) { viewStore in
          /* MyView */
      }
}
```

**Reply by mbrandonw** [reply]
@erikolsson Be sure to read our [performance article](https://pointfreeco.github.io/swift-composable-architecture/main/documentation/composablearchitecture/performance#View-stores) where we explain why it is generally not a good idea to observe all of the state in your store, especially at the root level of the application.

**Reply by erikolsson** [reply]
> @erikolsson Be sure to read our [performance article](https://pointfreeco.github.io/swift-composable-architecture/main/documentation/composablearchitecture/performance#View-stores) where we explain why it is generally not a good idea to observe all of the state in your store, especially at the root level of the application.

That code was just illustrative, I was observing only the necessary properties. Let me see if I can put together a test case. I had a problem where a .sheet -> navigation view -> alert would crash without a clear stack trace when dismissing.

More than anything, this felt like a bug in SwiftUI.

**Comment by tgrapperon** [comment]
Does making the root app action `indirect` fix your issue?

**Reply by jaanussiim** [reply]
From my side - I have since added/changed actions and, if I understand you correctly, I may have fixed it in the app by accident?

**Reply by tgrapperon** [reply]
If this is https://github.com/pointfreeco/swift-case-paths/issues/71 kicking in, it seems indeed that having more actions fixes the issue. Most likely because the enum's layout then falls back again into something accounted by the extraction's algorithm, and it succeeds again. I didn't check the `indirect` trick at the time, but it seems to have the same effect. On a two enums cases, it seems enough to make the root enum indirect, so I'm wondering if it propagates upward and if making the root action from the app `indirect` is fixing the whole app, even if the issue occurs at a deeper level.
Of course, this is only a workaround, and I can't guarantee it will always work. The solution would likely be to handle the situation explicitly, but it requires runtime/low-level knowledge that I don't have.

**Reply by PorterHoskins** [reply]
These workarounds don't work for me. I am also never using an enum with a single value.

**Reply by PorterHoskins** [reply]
I don't think `.debug()` works in `Release`. If `.debug` is supposed to work in `Release` then it is not getting the action because nothing prints in the console.

**Reply by tgrapperon** [reply]
I thought `.debug()` now worked in `Release` because I tried on `CaseStudies`, but I updated the `Test` scheme instead of the `Run` one. Sorry about that.

**Comment by stephencelis** [comment]
@PorterHoskins Since it seems like your bug is slightly different than @jaanussiim's, is there a chance you can provide a repro for us to mess around with?

**Reply by PorterHoskins** [reply]
I emailed you guys.

**Comment by PorterHoskins** [comment]
Exciting news. After boiling my massive project down to it's essence I figured out how to reproduce it 🎉

The essence of it is chain scoping a store is not working right. Maybe this is an anti-pattern so let me know if I am doing something wrong.

I am running in Release from Xcode 14.0 on an iPad running 15.6.1 and on an iPhone 12 running 16.0.

```
// Doesn't Work
ContentView(
     store: store
           .scope(state: \.reducer2, action: ReducerFeature.Action.reducer2)
           .scope(state: \.reducer3, action: Reducer2Feature.Action.reducer3)
)
            
 // Works
ContentView(
      store: reducer2Store
            .scope(state: \.reducer3, action: Reducer2Feature.Action.reducer3)
)
```

Full demo:

[ViewStoreUpdates.swift.zip](https://github.com/pointfreeco/swift-composable-architecture/files/9622458/ViewStoreUpdates.swift.zip)

**Reply by tgrapperon** [reply]
@PorterHoskins Indeed, if this is not the usual way, it should still work when chained like this.
@mbrandonw, @stephencelis, @iampatbrown, it's very late for me, but it's the `childStore` of `reducer2` created at l.648 that is not retained when scoping twice. It is not an existential issue, but most likely some oversight with `ScopedReducer`'s implementation. Should we keep references to the parents into it when we rescope?
Edit: [Something like this](https://github.com/pointfreeco/swift-composable-architecture/compare/protocol-beta...tgrapperon:swift-composable-architecture:protocol-beta-scoped-parents?expand=1)?

**Reply by mbrandonw** [reply]
Great find @PorterHoskins! That is very helpful.

@tgrapperon your fix makes sense, I'm just trying to get a repro in a test so that we can have some coverage on this. It's making me realize our tests aren't really set up to be easily run in release mode. Thinking we either need to revamp our tests to separate the stuff that can run in release from the stuff that must run in debug, or make a bunch of internal stuff public+underscored so that we can use it without `@testable import`.

**Reply by mbrandonw** [reply]
Ok, was able to get a failing test, and @tgrapperon's patch does fix it:

```swift
final class StoreTests: XCTestCase {
  var cancellables: Set<AnyCancellable> = []

  func testMultipleScopes() {
    let expectation = self.expectation(description: "")

    enum Action { case tap, response }
    let store = Store(
      initialState: false,
      reducer: Reduce<Bool, Action> { state, action in
        switch action {
        case .tap:
          state = false
          return .task { .response }
        case .response:
          state = true
          return .fireAndForget {
            expectation.fulfill()
          }
        }
      }
    )
    let viewStore = ViewStore(store.scope(state: { $0 }).scope(state: { $0 }))

    var values: [Bool] = []
    viewStore.publisher
      .sink(receiveValue: { values.append($0 )})
      .store(in: &self.cancellables)

    XCTAssertEqual(values, [false])
    viewStore.send(.tap)
    self.wait(for: [expectation], timeout: 1)
    XCTAssertEqual(values, [false, true])
  }
}
```

Interestingly it requires an effect to reproduce. Just sending actions and observing the state changes works fine. It's only when an action is sent from an effect do you see a problem.

**Reply by tgrapperon** [reply]
@mbrandonw This was a big chunk of very central logic that wasn't tested. Do you know if the reason that made reserving this path to release builds is still there now that 5.7 is official? 
Of course, release-mode tests would be nice, at least for the core library. Another approach would be to have a few simple testing helpers built on TCA's side. A quick check shows that no so many of them are required. This would probably make them more difficult to reach than simple underscores.

**Reply by mbrandonw** [reply]
> @mbrandonw This was a big chunk of very central logic that wasn't tested. Do you know if the reason that made reserving this path to release builds is still there now that 5.7 is official?

We are still supporting Swift <5.7 for a bit longer. We've even got some special reducer builder logic in place to make it more usable for people who can't use `some ReducerProtocol<State, Action>` yet.

I think to get us moving forward I'm just going to add this test to the main test suite, and fence all other tests in `#if DEBUG`, and in CI we will start running the test suite in both release and debug. And then we can slowly start chipping away making more and more tests run in both release and debug.

**Reply by tgrapperon** [reply]
Sorry, my question wasn't clear. As I get it, the `any Reducer<State, Action>` was reserved for release builds because of an issue in `DEBUG` builds. As this change was made during the beta, I was wondering if the issue responsible for the `!DEBUG` part of the check was still present in the official 5.7 release of Swift. I've checked on a branch without this additional check, and I didn't spot anything obvious.

**Reply by mbrandonw** [reply]
Ah yes yes, I see. Yeah I'm having trouble remembering what was wrong with `any Reducer` in debug. I know the `ReducerProtocolOf` typealias broke the editor... but what did `any Reducer` do? @stephencelis do you remember?

**Reply by tgrapperon** [reply]
As an aside, I've just checked the `ReducerProtocolOf` editor issue (https://github.com/apple/swift/issues/60550). This issue is present in Xcode 14.0, but it seems fixed in Xcode 14.1 beta!

**Reply by mbrandonw** [reply]
Fantastic news!

**Reply by mbrandonw** [reply]
@tgrapperon We just dropped the release-restriction on using `any ReducerProtocol` so please let us know if you see anything funky that might be able to be tracked back to that change.

**Reply by PorterHoskins** [reply]
Looks good on my end

**Comment by PorterHoskins** [comment]
@mbrandonw is there a chance `rescope` has this same issue? I'm seeing the same behavior for `ForEachStore` after the `parentStores` fix.

**Reply by mbrandonw** [reply]
Huh, the `parentStores` fix was specifically for `rescope`. Do you happen to have a repro I can play with?

**Reply by PorterHoskins** [reply]
Nvm I found it. It is not TCA causing the issue.

**Reply by mbrandonw** [reply]
😮‍💨 glad you figured it out!
